design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,485 @@
+
+We found the following to be useful in WhirlyGlobe-Maply:
+- Allocate a Buffer of a certain size with or without an NSData object to be copied in
+- Allocate a Texture Buffer of a certain size with or without an NSData object to be copied in
+- Return a handle to describe what was allocated.  If there's additional information, like a Heap, this is where to stash that.
+- If a given object is too large for the pre-allocated Heaps, just create it on its own
+- For a Drawable, reference the Heap and pass in the buffer offset as needed
+
+In the end, heaps are heaps and these are fairly standard
+
+#### Benefits
+
+Managing Metal memory with Heaps is more efficient than individual buffers.  Buffer management tends to dominate the low level wiring of Drawables.  For each frame, much of what you're doing on the CPU is pointing a drawing command at its buffers.  With Heaps, you still do that, but much of the checking goes away.  You make the appropriate Heaps available and then the driver is largely just dealing with offsets.
+
+It's fast and it's a key step to getting as much work as possible off the main CPU.
+
+### Indirect Encoding Implementation
+
+_Addresses Core Functionality [#7](#core)._
+
+The implementation up to this point is Direct.  For each frame we fill out the same command buffers.  Sure, the data buffers are reused, but the command buffers are created, filled in, and handed over to Metal each and every frame.  This is not dissimilar to OpenGL, in a way, but still more efficient.
+
+Any time you're doing the same job over and over again, the question is:  Can you not?  It turns out you can... not... do the work.
+
+![Direct vs Indirect version of the renderer](resources/figs-tmp-indirect.jpg)
+
+The trick with indirect rendering in Metal is to separate the parts that change a lot from the parts that don't change very often, or ever.  Uniforms for map position, lighting, etc may change every frame.  So you update them every frame.  But the geometry for a tile isn't going to.  Most of the uniforms for a set of roads won't either, except some might with data driven styling.
+
+There are a couple of ways to handle this.  One is to maintain a list of redundant buffers for anything that may change, one for each frame in flight.  So, for example, you may have 5 buffers for the core uniforms (matrices, lighting, etc) and you iterate over them for 5 concurrent frames.
+
+The approach we used in WhirlyGlobe-Maply was a bit more complex, but used less memory.  We set up each changeable uniform buffer for a given frame and then copied it into place using a Blit Encoder.  That's basically a construct used to copy memory around on the GPU.
+
+For Drawables, we looked for updates to any of their key values (texture IDs, colors, etc) and blitted those into place as needed.
```

Comment:
Would you go as far as leveraging compute shaders to build your indirect buffers? What might that look like, I wonder? And how much value would there be in going heavily indirect with async GPU compute?

Reply:
We considered the first.  I had envisioned that for "expensive" operations we might do the calculations once in a compute shader and leave them in a memory buffer for other shaders to pick up.  There were just never expensive enough operations that need that approach.  I still like the idea.

I suppose it's not dissimilar from our approach to particles.  We do have a calculation shader for that, which saves its results.  Then another render shader that picks it up later to draw.  So conceptually I suppose we have already done it.  Worked great.

For MapLibre Native, as to the second question, what is it that would demand that sort of approach?  Dynamic reprojection and resampling of surfaces on the fly?  Intriguing, but the toolkit doesn't do that now.  Perhaps in the future.

Reply:
The only times I've considered it are for computing and caching expensive and/or frequent vertex buffer updates. An example, dynamic tessellated displacement mapped terrain. Instead of using the hardware tessellation as part of your shader pipeline (hull/domain / tesc/tese) you tessellate in a compute shader and store that in a buffer to draw indirectly.

The above gets more appealing when you have something like a Z pre-pass, using the hardware tessellation pipeline means running your tessellation shader twice (ouch!). An alternative is transform feedback, but I don't see that used as much anymore since compute is so much more flexible.

Applied to MapLibre, it would depend on how complex the existing triangulation routines are for various feature types. The most expensive is probably line layer, but I'm not sure it's at the point where it's worth it to go through all that. This may be something that becomes appealing in the future if 3D extensions take off though (see maplibre-js 3D terrain).

Reply:
I doubt it's worth it for anything at the moment.  Widened lines just aren't going to produce enough volume for their vertex shaders to overwhelm the system.

I have occasionally mused about just doing things like tessellation of lines in its own pass because it would be cleaner.  There's a certain appeal to not duplicating work and just laying out each unique stage in its own pass.  But that's one of the weirder tradeoffs we make in these things.  A pass is costly and it's just usually cheaper to duplicate work.

Even so, I'd like this kind of thing to be easier to do in MapLibre Native.  If someone wants to try it out, then they need to get after the LineBuilder to emit a calculation pass and then a rendering pass.  And ideally that's all (as if that's not enough work).  But if we can avoid them having to tear down half the renderer to try something like this, we've succeeded, at least in part.

Reply:
Yes.  There are cases where that might be useful, such as the ones we were discussing.  I don't think we would use them in the first implementation, but we would want to make it possible for others to construct them from components.  Again, we don't have to build any, but just keep them in mind when building the Metal support.